app/src/main/java/com/example/criminalintentapp/CrimeListFragment.kt

```diff
@@ -0,0 +1,104 @@
+package com.example.criminalintentapp
+
+import android.os.Bundle
+import android.util.Log
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.TextView
+import android.widget.Toast
+import androidx.fragment.app.Fragment
+import androidx.lifecycle.ViewModelProvider
+import androidx.recyclerview.widget.LinearLayoutManager
+import androidx.recyclerview.widget.RecyclerView
+
+class CrimeListFragment : Fragment() {
+
+    private lateinit var crimeRecyclerView: RecyclerView
+    private var adapter: CrimeAdapter? = null
+
+    private val crimeListViewModel: CrimeListViewModel by lazy {
+        ViewModelProvider(this).get(CrimeListViewModel::class.java)
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        Log.d(TAG, "Total crimes: ${crimeListViewModel.crimes.size}")
+    }
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        val view = inflater.inflate(R.layout.fragment_crime_list, container, false)
+
+        bindViews(view)
+
+        crimeRecyclerView.layoutManager = LinearLayoutManager(context)
+
+        updateUI()
+
+        return view
+    }
+
+    private fun bindViews(view: View) {
+        crimeRecyclerView = view.findViewById(R.id.crime_recycler_view) as RecyclerView
+    }
+
+    private fun updateUI() {
+        val crimes = crimeListViewModel.crimes
+        adapter = CrimeAdapter(crimes)
+        crimeRecyclerView.adapter = adapter
+    }
+
+    private inner class CrimeHolder(view: View) :
+        RecyclerView.ViewHolder(view),
+        View.OnClickListener {
+
+        private lateinit var crime: Crime
+
+        private val titleTextView: TextView = itemView.findViewById(R.id.crime_title)
+        private val dateTextView: TextView = itemView.findViewById(R.id.crime_date)
+
+        init {
+            itemView.setOnClickListener(this)
+        }
+
+        fun bind(crime: Crime) {
+            this.crime = crime
+            titleTextView.text = this.crime.title
+            dateTextView.text = this.crime.date.toString()
+        }
+
+        override fun onClick(v: View?) {
+            Toast.makeText(context, "${crime.title} pressed", Toast.LENGTH_SHORT).show()
+        }
+    }
+
+    private inner class CrimeAdapter(var crimes: List<Crime>) :
```

Comment:
In case the book does not mention this aspect, you will see that in most projects the adapters are found in a separate file. You should create a CrimeAdapter.kt where you put the code for the adapter and the viewholder(they are tightly connected so it makes sense to have them in the same file)